File: src/main/java/cn/keevol/keenotes/mobilefx/NotePosterRenderer.java
package cn.keevol.keenotes.mobilefx;

import cn.keevol.keenotes.mobilefx.utils.DateTimeUtil;
import javafx.scene.image.Image;

import javax.imageio.ImageIO;
import java.awt.AlphaComposite;
import java.awt.BasicStroke;
import java.awt.Color;
import java.awt.Composite;
import java.awt.Font;
import java.awt.FontFormatException;
import java.awt.FontMetrics;
import java.awt.Graphics2D;
import java.awt.LinearGradientPaint;
import java.awt.RenderingHints;
import java.awt.Shape;
import java.awt.geom.RoundRectangle2D;
import java.awt.image.BufferedImage;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;

public final class NotePosterRenderer {
    public static final int EXPORT_WIDTH = 1080;
    public static final double ASPECT_RATIO = 9.0 / 16.0;
    private static final String POSTER_FONT_RESOURCE = "/fonts/MiSans-Regular.ttf";
    private static final Font BASE_POSTER_FONT = loadBasePosterFont();

    private static final int PAPER_RGB = 0xFDFCFB;
    private static final Color INK_TEXT = new Color(26, 24, 22);
    private static final Color FOOTER_TEXT = new Color(26, 24, 22, 88);
    private static final Color BADGE_FILL = new Color(0, 0, 0, 12);
    private static final Color BADGE_STROKE = new Color(0, 0, 0, 14);
    private static final Color DIVIDER = new Color(0, 0, 0, 15);

    private NotePosterRenderer() {
    }

    public static BufferedImage renderPosterImage(
            LocalCacheService.NoteData noteData,
            String hiddenMessage,
            PosterInkTheme inkTheme
    ) throws IOException {
        return renderPosterImage(noteData.content, formatPosterDate(noteData.createdAt), hiddenMessage, inkTheme, EXPORT_WIDTH);
    }

    public static BufferedImage renderPosterImage(
            String noteContent,
            String posterDate,
            String hiddenMessage,
            PosterInkTheme inkTheme,
            int width
    ) throws IOException {
        String content = noteContent == null ? "" : noteContent;
        String author = trimToNull(hiddenMessage);
        int contentLength = content.codePointCount(0, content.length());
        double scale = width / 390.0;

        int cardPadding = scaled(contentLength > 700 ? 28 : 34, scale);
        int contentVerticalPadding = scaled(resolveContentVerticalPadding(contentLength), scale);
        int footerVerticalPadding = scaled(contentLength > 700 ? 18 : 22, scale);
        int contentFontSize = scaled(resolveContentFontSize(contentLength), scale);
        int contentLineSpacing = scaled(resolveContentLineSpacing(contentLength), scale);
        int footerFontSize = scaled(11, scale);

        Font contentFont = posterFont(Font.BOLD, contentFontSize);
        Font footerFont = posterFont(Font.PLAIN, footerFontSize);
        Font badgeFont = posterFont(Font.BOLD, footerFontSize);

        BufferedImage measureImage = new BufferedImage(1, 1, BufferedImage.TYPE_INT_ARGB);
        Graphics2D measureGraphics = measureImage.createGraphics();
        configureGraphics(measureGraphics);
        measureGraphics.setFont(contentFont);
        FontMetrics contentMetrics = measureGraphics.getFontMetrics();
        List<String> lines = wrapText(content, contentMetrics, width - cardPadding * 2);
        int lineAdvance = contentMetrics.getHeight() + contentLineSpacing;
        int textHeight = lines.isEmpty() ? contentMetrics.getHeight() : lines.size() * lineAdvance - contentLineSpacing;

        measureGraphics.setFont(footerFont);
        int footerTextHeight = measureGraphics.getFontMetrics().getHeight();
        measureGraphics.dispose();

        int footerHeight = footerTextHeight + footerVerticalPadding * 2;
        int dividerHeight = Math.max(1, scaled(1, scale));
        int minimumHeight = exportHeightFor(width);
        int requiredHeight = contentVerticalPadding * 2 + textHeight + dividerHeight + footerHeight;
        int height = Math.max(minimumHeight, requiredHeight);
        int extraVerticalSpace = height - requiredHeight;
        int textY = contentVerticalPadding + extraVerticalSpace / 2;
        int footerY = height - footerHeight;

        BufferedImage image = new BufferedImage(width, height, BufferedImage.TYPE_INT_ARGB);
        Graphics2D g = image.createGraphics();
        configureGraphics(g);

        int cornerRadius = scaled(32, scale);
        Shape posterShape = new RoundRectangle2D.Double(0, 0, width, height, cornerRadius, cornerRadius);
        g.setClip(posterShape);
        g.setColor(new Color(PAPER_RGB));
        g.fillRect(0, 0, width, height);

        paintPaperGrain(g, width, height, scale);
        paintInkTheme(g, inkTheme, contentLength, width, height);
        paintVignette(g, width, height);
        paintContent(g, lines, contentFont, contentMetrics, cardPadding, textY, lineAdvance);
        paintFooter(g, author, posterDate, footerFont, badgeFont, cardPadding, footerY, footerHeight, footerVerticalPadding, width);

        g.setColor(DIVIDER);
        g.fillRect(0, footerY, width, dividerHeight);

        g.setClip(null);
        g.setStroke(new BasicStroke(Math.max(1f, (float) (0.8 * scale))));
        g.setColor(new Color(255, 255, 255, 210));
        g.draw(posterShape);
        g.dispose();
        return image;
    }

    public static Image toFxImage(BufferedImage image) throws IOException {
        ByteArrayOutputStream output = new ByteArrayOutputStream();
        ImageIO.write(image, "png", output);
        return new Image(new ByteArrayInputStream(output.toByteArray()));
    }

    public static void writePng(BufferedImage image, File outputFile) throws IOException {
        ImageIO.write(image, "png", outputFile);
    }

    public static int exportHeightFor(int width) {
        return (int) Math.ceil(width / ASPECT_RATIO);
    }

    public static String formatPosterDate(String createdAt) {
        String local = DateTimeUtil.utcToLocalDisplay(createdAt);
        if (local != null && local.length() >= 10) {
            return local.substring(0, 10);
        }
        return local == null || local.isBlank() ? "" : local;
    }

    private static int resolveContentFontSize(int contentLength) {
        if (contentLength > 900) {
            return 16;
        }
        if (contentLength > 500) {
            return 18;
        }
        if (contentLength > 220) {
            return 20;
        }
        return 24;
    }

    private static int resolveContentLineSpacing(int contentLength) {
        if (contentLength > 700) {
            return 8;
        }
        if (contentLength > 300) {
            return 10;
        }
        return 12;
    }

    private static int resolveContentVerticalPadding(int contentLength) {
        if (contentLength > 900) {
            return 30;
        }
        if (contentLength > 500) {
            return 38;
        }
        return 52;
    }

    private static void configureGraphics(Graphics2D g) {
        g.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
        g.setRenderingHint(RenderingHints.KEY_TEXT_ANTIALIASING, RenderingHints.VALUE_TEXT_ANTIALIAS_ON);
        g.setRenderingHint(RenderingHints.KEY_RENDERING, RenderingHints.VALUE_RENDER_QUALITY);
        g.setRenderingHint(RenderingHints.KEY_INTERPOLATION, RenderingHints.VALUE_INTERPOLATION_BICUBIC);
    }

    private static Font posterFont(int style, int size) {
        return BASE_POSTER_FONT.deriveFont(style, (float) size);
    }

    private static Font loadBasePosterFont() {
        try (InputStream input = NotePosterRenderer.class.getResourceAsStream(POSTER_FONT_RESOURCE)) {
            if (input != null) {
                return Font.createFont(Font.TRUETYPE_FONT, input);
            }
        } catch (FontFormatException | IOException e) {
            // Fall through to platform sans-serif if the bundled font cannot be loaded.
        }
        return new Font(Font.SANS_SERIF, Font.PLAIN, 12);
    }

    private static void paintPaperGrain(Graphics2D g, int width, int height, double scale) {
        int step = Math.max(1, scaled(7, scale));
        for (int y = 0; y < height; y += step) {
            for (int x = 0; x < width; x += step) {
                int seed = Math.round(x * 31 + y * 17);
                int alpha = seed % 5 == 0 ? 7 : 4;
                g.setColor(new Color(0, 0, 0, alpha));
                g.fillRect(x, y, Math.max(1, scaled(1, scale)), Math.max(1, scaled(1, scale)));
            }
        }
    }

    private static void paintInkTheme(Graphics2D g, PosterInkTheme theme, int contentLength, int width, int height) throws IOException {
        try (InputStream input = NotePosterRenderer.class.getResourceAsStream(theme.getResourcePath())) {
            if (input == null) {
                return;
            }
            BufferedImage ink = ImageIO.read(input);
            if (ink == null) {
                return;
            }
            Composite previousComposite = g.getComposite();
            g.setComposite(AlphaComposite.getInstance(AlphaComposite.SRC_OVER, theme.opacityFor(contentLength)));
            int inkHeight = exportHeightFor(width);
            int y = height - inkHeight;
            g.drawImage(ink, 0, y, width, inkHeight, null);
            g.setComposite(previousComposite);
        }
    }

    private static void paintVignette(Graphics2D g, int width, int height) {
        LinearGradientPaint edgeShade = new LinearGradientPaint(
                0, 0, width, height,
                new float[]{0f, 0.5f, 1f},
                new Color[]{new Color(255, 255, 255, 0), new Color(255, 255, 255, 0), new Color(0, 0, 0, 8)}
        );
        g.setPaint(edgeShade);
        g.fillRect(0, 0, width, height);
    }

    private static void paintContent(Graphics2D g, List<String> lines, Font font, FontMetrics metrics, int x, int y, int lineAdvance) {
        g.setFont(font);
        g.setColor(INK_TEXT);
        int baseline = y + metrics.getAscent();
        if (lines.isEmpty()) {
            return;
        }
        for (String line : lines) {
            g.drawString(line, x, baseline);
            baseline += lineAdvance;
        }
    }

    private static void paintFooter(
            Graphics2D g,
            String author,
            String posterDate,
            Font footerFont,
            Font badgeFont,
            int cardPadding,
            int footerY,
            int footerHeight,
            int footerVerticalPadding,
            int width
    ) {
        String leftText = author == null ? posterDate : author + " · " + posterDate;
        g.setFont(footerFont);
        FontMetrics footerMetrics = g.getFontMetrics();
        int baseline = footerY + footerVerticalPadding + footerMetrics.getAscent();

        g.setFont(badgeFont);
        FontMetrics badgeMetrics = g.getFontMetrics();
        String badgeText = "KeeNotes";
        int badgeHorizontalPadding = Math.max(10, badgeMetrics.getHeight() / 2);
        int badgeVerticalPadding = Math.max(5, badgeMetrics.getHeight() / 4);
        int badgeWidth = badgeMetrics.stringWidth(badgeText) + badgeHorizontalPadding * 2;
        int badgeHeight = badgeMetrics.getHeight() + badgeVerticalPadding * 2;
        int badgeX = width - cardPadding - badgeWidth;
        int badgeY = footerY + (footerHeight - badgeHeight) / 2;
        int badgeArc = badgeHeight;

        g.setColor(BADGE_FILL);
        g.fillRoundRect(badgeX, badgeY, badgeWidth, badgeHeight, badgeArc, badgeArc);
        g.setColor(BADGE_STROKE);
        g.drawRoundRect(badgeX, badgeY, badgeWidth, badgeHeight, badgeArc, badgeArc);
        g.setColor(new Color(26, 24, 22, 148));
        g.drawString(badgeText, badgeX + badgeHorizontalPadding, badgeY + badgeVerticalPadding + badgeMetrics.getAscent());

        g.setFont(footerFont);
        g.setColor(FOOTER_TEXT);
        int availableLeftWidth = Math.max(0, badgeX - cardPadding - Math.max(12, footerMetrics.getHeight()));
        g.drawString(ellipsize(leftText, footerMetrics, availableLeftWidth), cardPadding, baseline);
    }

    private static List<String> wrapText(String text, FontMetrics metrics, int maxWidth) {
        List<String> lines = new ArrayList<>();
        String normalized = text.replace("\r\n", "\n").replace('\r', '\n');
        String[] paragraphs = normalized.split("\n", -1);
        for (int i = 0; i < paragraphs.length; i++) {
            wrapParagraph(paragraphs[i], metrics, maxWidth, lines);
            if (i < paragraphs.length - 1) {
                lines.add("");
            }
        }
        while (!lines.isEmpty() && lines.get(lines.size() - 1).isEmpty()) {
            lines.remove(lines.size() - 1);
        }
        return lines;
    }

    private static void wrapParagraph(String paragraph, FontMetrics metrics, int maxWidth, List<String> lines) {
        if (paragraph.isEmpty()) {
            return;
        }

        StringBuilder current = new StringBuilder();
        for (int offset = 0; offset < paragraph.length(); ) {
            int codePoint = paragraph.codePointAt(offset);
            String next = new String(Character.toChars(codePoint));
            String candidate = current + next;
            if (metrics.stringWidth(candidate) <= maxWidth || current.isEmpty()) {
                current.append(next);
            } else {
                lines.add(current.toString());
                current.setLength(0);
                current.append(next);
            }
            offset += Character.charCount(codePoint);
        }

        if (!current.isEmpty()) {
            lines.add(current.toString());
        }
    }

    private static String ellipsize(String text, FontMetrics metrics, int maxWidth) {
        if (text == null || metrics.stringWidth(text) <= maxWidth) {
            return text == null ? "" : text;
        }
        String ellipsis = "...";
        StringBuilder result = new StringBuilder(text);
        while (!result.isEmpty() && metrics.stringWidth(result + ellipsis) > maxWidth) {
            result.setLength(result.length() - 1);
        }
        return result + ellipsis;
    }

    private static String trimToNull(String value) {
        if (value == null) {
            return null;
        }
        String trimmed = value.trim();
        return trimmed.isEmpty() ? null : trimmed;
    }

    private static int scaled(double value, double scale) {
        return Math.max(1, (int) Math.round(value * scale));
    }
}

File: keenotes-android/app/src/main/java/cn/keevol/keenotes/share/NotePosterRenderer.kt
package cn.keevol.keenotes.share

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.BlendMode
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.LinearGradient
import android.graphics.Paint
import android.graphics.Path
import android.graphics.Rect
import android.graphics.RectF
import android.graphics.Shader
import android.graphics.Typeface
import android.text.Layout
import android.text.StaticLayout
import android.text.TextPaint
import android.text.TextUtils
import cn.keevol.keenotes.data.entity.Note
import cn.keevol.keenotes.util.DateTimeUtil
import kotlin.math.ceil
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

object NotePosterRenderer {
    const val EXPORT_WIDTH = 1080
    const val VIDEO_WIDTH = 1080
    const val VIDEO_HEIGHT = 1920
    const val ASPECT_RATIO = 9f / 16f

    private val paperColor = Color.rgb(0xFD, 0xFC, 0xFB)
    private val inkText = Color.rgb(26, 24, 22)
    private val footerText = Color.argb(88, 26, 24, 22)
    private val badgeFill = Color.argb(12, 0, 0, 0)
    private val badgeStroke = Color.argb(14, 0, 0, 0)
    private val divider = Color.argb(15, 0, 0, 0)

    fun renderPosterImage(
        context: Context,
        note: Note,
        hiddenMessage: String,
        inkTheme: PosterInkTheme,
        width: Int = EXPORT_WIDTH
    ): Bitmap {
        return renderPosterImage(
            context = context,
            noteContent = note.content,
            posterDate = formatPosterDate(note.createdAt),
            hiddenMessage = hiddenMessage,
            inkTheme = inkTheme,
            width = width
        )
    }

    fun renderPosterImage(
        context: Context,
        noteContent: String,
        posterDate: String,
        hiddenMessage: String,
        inkTheme: PosterInkTheme,
        width: Int = EXPORT_WIDTH
    ): Bitmap {
        val content = noteContent
        val author = hiddenMessage.trim().takeIf { it.isNotEmpty() }
        val contentLength = content.codePointCount(0, content.length)
        val scale = width / 390f

        val cardPadding = scaled(if (contentLength > 700) 28 else 34, scale)
        val contentVerticalPadding = scaled(resolveContentVerticalPadding(contentLength), scale)
        val footerVerticalPadding = scaled(if (contentLength > 700) 18 else 22, scale)
        val contentFontSize = scaled(resolveContentFontSize(contentLength), scale).toFloat()
        val contentLineSpacing = scaled(resolveContentLineSpacing(contentLength), scale).toFloat()
        val footerFontSize = scaled(11, scale).toFloat()

        val contentPaint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
            color = inkText
            textSize = contentFontSize
            typeface = Typeface.create(Typeface.SERIF, Typeface.BOLD)
        }
        val footerPaint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
            color = footerText
            textSize = footerFontSize
            typeface = Typeface.create(Typeface.SANS_SERIF, Typeface.NORMAL)
        }
        val badgePaint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
            color = Color.argb(148, 26, 24, 22)
            textSize = footerFontSize
            typeface = Typeface.create(Typeface.SANS_SERIF, Typeface.BOLD)
        }

        val textWidth = width - cardPadding * 2
        val layout = buildLayout(content, contentPaint, textWidth, contentLineSpacing)
        val textHeight = if (content.isEmpty()) {
            val metrics = contentPaint.fontMetricsInt
            metrics.descent - metrics.ascent
        } else {
            layout.height
        }
        val footerTextHeight = run {
            val metrics = footerPaint.fontMetricsInt
            metrics.descent - metrics.ascent
        }

        val footerHeight = footerTextHeight + footerVerticalPadding * 2
        val dividerHeight = max(1, scaled(1, scale))
        val minimumHeight = exportHeightFor(width)
        val requiredHeight = contentVerticalPadding * 2 + textHeight + dividerHeight + footerHeight
        val height = max(minimumHeight, requiredHeight)
        val extraVerticalSpace = height - requiredHeight
        val textY = contentVerticalPadding + extraVerticalSpace / 2
        val footerY = height - footerHeight

        val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        val cornerRadius = scaled(32, scale).toFloat()
        val posterRect = RectF(0f, 0f, width.toFloat(), height.toFloat())
        val posterPath = Path().apply {
            addRoundRect(posterRect, cornerRadius, cornerRadius, Path.Direction.CW)
        }

        canvas.save()
        canvas.clipPath(posterPath)
        canvas.drawColor(paperColor)
        paintPaperGrain(canvas, width, height, scale)
        paintInkTheme(context, canvas, inkTheme, contentLength, width, height)
        paintVignette(canvas, width, height)
        paintContent(canvas, layout, content, cardPadding, textY)

        val dividerPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { color = divider }
        canvas.drawRect(0f, footerY.toFloat(), width.toFloat(), (footerY + dividerHeight).toFloat(), dividerPaint)
        paintFooter(canvas, author, posterDate, footerPaint, badgePaint, cardPadding, footerY, footerVerticalPadding, width)
        canvas.restore()

        val outlinePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = Color.argb(210, 255, 255, 255)
            style = Paint.Style.STROKE
            strokeWidth = max(1f, 0.8f * scale)
        }
        canvas.drawRoundRect(posterRect, cornerRadius, cornerRadius, outlinePaint)

        return bitmap
    }

    fun renderVideoFrame(posterImage: Bitmap): Bitmap {
        val frame = Bitmap.createBitmap(VIDEO_WIDTH, VIDEO_HEIGHT, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(frame)
        val frameRect = RectF(0f, 0f, VIDEO_WIDTH.toFloat(), VIDEO_HEIGHT.toFloat())
        canvas.drawColor(paperColor)

        val smallBackground = Bitmap.createBitmap(270, 480, Bitmap.Config.ARGB_8888)
        Canvas(smallBackground).apply {
            drawColor(paperColor)
            drawBitmapCover(posterImage, RectF(0f, 0f, width.toFloat(), height.toFloat()), Paint(Paint.ANTI_ALIAS_FLAG))
        }
        val blurred = boxBlur(smallBackground, 10)
        canvas.drawBitmap(blurred, null, frameRect, Paint(Paint.ANTI_ALIAS_FLAG).apply {
            isFilterBitmap = true
        })
        smallBackground.recycle()
        blurred.recycle()

        val scale = min(VIDEO_WIDTH / posterImage.width.toFloat(), VIDEO_HEIGHT / posterImage.height.toFloat())
        val foregroundWidth = posterImage.width * scale
        val foregroundHeight = posterImage.height * scale
        val left = (VIDEO_WIDTH - foregroundWidth) / 2f
        val top = (VIDEO_HEIGHT - foregroundHeight) / 2f
        canvas.drawBitmap(
            posterImage,
            null,
            RectF(left, top, left + foregroundWidth, top + foregroundHeight),
            Paint(Paint.ANTI_ALIAS_FLAG).apply { isFilterBitmap = true }
        )
        return frame
    }

    fun formatPosterDate(createdAt: String): String {
        val local = DateTimeUtil.utcToLocalDisplay(createdAt)
        return if (local.length >= 10) local.substring(0, 10) else local
    }

    fun exportHeightFor(width: Int): Int = ceil(width / ASPECT_RATIO).toInt()

    private fun buildLayout(text: String, paint: TextPaint, width: Int, lineSpacing: Float): StaticLayout {
        return StaticLayout.Builder.obtain(text, 0, text.length, paint, width)
            .setAlignment(Layout.Alignment.ALIGN_NORMAL)
            .setIncludePad(false)
            .setLineSpacing(lineSpacing, 1f)
            .build()
    }

    private fun paintPaperGrain(canvas: Canvas, width: Int, height: Int, scale: Float) {
        val step = max(1, scaled(7, scale))
        val dotSize = max(1f, scaled(1, scale).toFloat())
        val paint = Paint()
        var y = 0
        while (y < height) {
            var x = 0
            while (x < width) {
                val seed = (x * 31 + y * 17).toFloat().roundToInt()
                val alpha = if (seed % 5 == 0) 7 else 4
                paint.color = Color.argb(alpha, 0, 0, 0)
                canvas.drawRect(x.toFloat(), y.toFloat(), x + dotSize, y + dotSize, paint)
                x += step
            }
            y += step
        }
    }

    private fun paintInkTheme(
        context: Context,
        canvas: Canvas,
        theme: PosterInkTheme,
        contentLength: Int,
        width: Int,
        height: Int
    ) {
        val ink = BitmapFactory.decodeResource(context.resources, theme.drawableRes) ?: return
        val inkHeight = exportHeightFor(width)
        val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            alpha = theme.alphaFor(contentLength)
            blendMode = BlendMode.MULTIPLY
            isFilterBitmap = true
        }
        try {
            canvas.drawBitmap(
                ink,
                null,
                Rect(0, height - inkHeight, width, height),
                paint
            )
        } finally {
            ink.recycle()
        }
    }

    private fun paintVignette(canvas: Canvas, width: Int, height: Int) {
        val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            shader = LinearGradient(
                0f,
                0f,
                width.toFloat(),
                height.toFloat(),
                intArrayOf(Color.TRANSPARENT, Color.TRANSPARENT, Color.argb(8, 0, 0, 0)),
                floatArrayOf(0f, 0.5f, 1f),
                Shader.TileMode.CLAMP
            )
        }
        canvas.drawRect(0f, 0f, width.toFloat(), height.toFloat(), paint)
    }

    private fun paintContent(canvas: Canvas, layout: StaticLayout, content: String, x: Int, y: Int) {
        if (content.isEmpty()) return
        canvas.save()
        canvas.translate(x.toFloat(), y.toFloat())
        layout.draw(canvas)
        canvas.restore()
    }

    private fun paintFooter(
        canvas: Canvas,
        author: String?,
        posterDate: String,
        footerPaint: TextPaint,
        badgePaint: TextPaint,
        cardPadding: Int,
        footerY: Int,
        footerVerticalPadding: Int,
        width: Int
    ) {
        val badgeText = "KeeNotes"
        val badgeMetrics = badgePaint.fontMetrics
        val badgeTextHeight = badgeMetrics.descent - badgeMetrics.ascent
        val badgeHorizontalPadding = max(10f, badgeTextHeight / 2f)
        val badgeVerticalPadding = max(5f, badgeTextHeight / 4f)
        val badgeWidth = badgePaint.measureText(badgeText) + badgeHorizontalPadding * 2f
        val badgeHeight = badgeTextHeight + badgeVerticalPadding * 2f
        val badgeX = width - cardPadding - badgeWidth
        val badgeY = footerY + footerVerticalPadding - badgeVerticalPadding
        val badgeRect = RectF(badgeX, badgeY, badgeX + badgeWidth, badgeY + badgeHeight)

        val badgeBackground = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = badgeFill
            style = Paint.Style.FILL
        }
        val badgeOutline = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = badgeStroke
            style = Paint.Style.STROKE
            strokeWidth = 1f
        }
        canvas.drawRoundRect(badgeRect, badgeHeight / 2f, badgeHeight / 2f, badgeBackground)
        canvas.drawRoundRect(badgeRect, badgeHeight / 2f, badgeHeight / 2f, badgeOutline)
        canvas.drawText(
            badgeText,
            badgeX + badgeHorizontalPadding,
            badgeY + badgeVerticalPadding - badgeMetrics.ascent,
            badgePaint
        )

        val leftText = if (author == null) posterDate else "$author · $posterDate"
        val availableWidth = (badgeX - cardPadding - 12f).coerceAtLeast(0f)
        val ellipsized = TextUtils.ellipsize(leftText, footerPaint, availableWidth, TextUtils.TruncateAt.END).toString()
        val footerMetrics = footerPaint.fontMetrics
        val baseline = footerY + footerVerticalPadding - footerMetrics.ascent
        canvas.drawText(ellipsized, cardPadding.toFloat(), baseline, footerPaint)
    }

    private fun Canvas.drawBitmapCover(bitmap: Bitmap, dst: RectF, paint: Paint) {
        val scale = max(dst.width() / bitmap.width.toFloat(), dst.height() / bitmap.height.toFloat())
        val scaledWidth = bitmap.width * scale
        val scaledHeight = bitmap.height * scale
        val left = dst.left + (dst.width() - scaledWidth) / 2f
        val top = dst.top + (dst.height() - scaledHeight) / 2f
        drawBitmap(bitmap, null, RectF(left, top, left + scaledWidth, top + scaledHeight), paint)
    }

    private fun boxBlur(source: Bitmap, radius: Int): Bitmap {
        if (radius <= 0) return source.copy(Bitmap.Config.ARGB_8888, false)
        val width = source.width
        val height = source.height
        val input = IntArray(width * height)
        val temp = IntArray(width * height)
        val output = IntArray(width * height)
        source.getPixels(input, 0, width, 0, 0, width, height)
        blurHorizontal(input, temp, width, height, radius)
        blurVertical(temp, output, width, height, radius)
        return Bitmap.createBitmap(output, width, height, Bitmap.Config.ARGB_8888)
    }

    private fun blurHorizontal(input: IntArray, output: IntArray, width: Int, height: Int, radius: Int) {
        val window = radius * 2 + 1
        for (y in 0 until height) {
            var a = 0
            var r = 0
            var g = 0
            var b = 0
            val row = y * width
            for (i in -radius..radius) {
                val c = input[row + i.coerceIn(0, width - 1)]
                a += Color.alpha(c)
                r += Color.red(c)
                g += Color.green(c)
                b += Color.blue(c)
            }
            for (x in 0 until width) {
                output[row + x] = Color.argb(a / window, r / window, g / window, b / window)
                val remove = input[row + (x - radius).coerceIn(0, width - 1)]
                val add = input[row + (x + radius + 1).coerceIn(0, width - 1)]
                a += Color.alpha(add) - Color.alpha(remove)
                r += Color.red(add) - Color.red(remove)
                g += Color.green(add) - Color.green(remove)
                b += Color.blue(add) - Color.blue(remove)
            }
        }
    }

    private fun blurVertical(input: IntArray, output: IntArray, width: Int, height: Int, radius: Int) {
        val window = radius * 2 + 1
        for (x in 0 until width) {
            var a = 0
            var r = 0
            var g = 0
            var b = 0
            for (i in -radius..radius) {
                val c = input[i.coerceIn(0, height - 1) * width + x]
                a += Color.alpha(c)
                r += Color.red(c)
                g += Color.green(c)
                b += Color.blue(c)
            }
            for (y in 0 until height) {
                output[y * width + x] = Color.argb(a / window, r / window, g / window, b / window)
                val remove = input[(y - radius).coerceIn(0, height - 1) * width + x]
                val add = input[(y + radius + 1).coerceIn(0, height - 1) * width + x]
                a += Color.alpha(add) - Color.alpha(remove)
                r += Color.red(add) - Color.red(remove)
                g += Color.green(add) - Color.green(remove)
                b += Color.blue(add) - Color.blue(remove)
            }
        }
    }

    private fun resolveContentFontSize(contentLength: Int): Int = when {
        contentLength > 900 -> 16
        contentLength > 500 -> 18
        contentLength > 220 -> 20
        else -> 24
    }

    private fun resolveContentLineSpacing(contentLength: Int): Int = when {
        contentLength > 700 -> 8
        contentLength > 300 -> 10
        else -> 12
    }

    private fun resolveContentVerticalPadding(contentLength: Int): Int = when {
        contentLength > 900 -> 30
        contentLength > 500 -> 38
        else -> 52
    }

    private fun scaled(value: Int, scale: Float): Int = max(1, (value * scale).roundToInt())
}
